spec.html

```diff
@@ -241,11 +244,36 @@ <h1>Static Semantics: HostGetSupportedImportAssertions ()</h1>
     <emu-note type=editor>The purpose of requiring the host to specify its supported import assertions, rather than passing all assertions to the host and letting it then choose which ones it wants to handle, is to ensure that unsupported assertions are handled in a consistent way across different hosts.</emu-note>
   </emu-clause>
 
+  <emu-clause id="sec-hostgetsupportedextramodulescripttypes" aoid="HostGetSupportedExtraModuleScriptTypes">
+    <h1>Static Semantics: HostGetSupportedExtraModuleScriptTypes ()</h1>
+    <p>
+      HostGetSupportedExtraModuleScriptTypes is a host-defined abstract operation that allows host environments to specify which types of module scripts they support in addition to Source Text Module Record.
+      Module script type assertions with unsupported values will cause an error.
+    </p>
+
+    <p>The implementation of HostGetSupportedExtraModuleScriptTypes must conform to the following requrements:</p>
+
+    <ul>
+      <li>It must return a List whose values are all StringValues, each indicating a supported module script type.</li>
+
+      <li>None of the string values may be *"js"* or *"javascript"*.<emu-note type="editor">It is assumed that Source Text Module Record must be supported.</emu-note></li>
```

Comment:
this seems like it prohibits hosts from making an optional “javascript” or “js” type, but doesn’t actually prevent them from supporting source text module records, or any other JavaScript-representing type under another name. What’s the benefit to this constraint?

Reply:
The intent is to enforce that neither of the following can be used to import a source text module record:
```js
import "./foo.js" assert { type: "js" };
import "./foo.js" assert { type: "javascript" };
```
There was some discussion in https://github.com/tc39/proposal-import-assertions/issues/49 about whether or not those should be permitted.  It seems to me that the decision of whether or not they work or not should be up to ECMAScript, not the individual hosts.  ECMAScript can choose to allow them or disallow them by editing whether they are restricted in this `<li>`, and/or adding them to the HostGetSupportedExtraModuleScriptTypes list if the host didn't add them.

On the other hand if we want the hosts to make that call then this item can just be deleted.

Reply:
The current text of this PR does not prohibit a host from doing:
```jsx
import "./foo.js" assert { type: "ecmascript" };
import "./foo.js" assert { type: "jscript" };
import "./foo.js" assert { type: "JS" };
import "./foo.js" assert { type: "JavaScript" };
```

In other words, if the goal is to prohibit hosts from allowing a type on source text-based modules, then this restriction, as written, doesn't achieve it - it only prohibits 2 of the infinite possible strings that could be used.

Reply:
I think we can only make that restriction in prose, right? Similar to how we have prose restrictions on job queuing. It seems reasonable to do that.

Reply:
So something like "None of the string values must be used for Source Text Module Records."

Reply:
That seems like it would achieve the goal.

However, is there an advantage to prohibiting an _optional_ type value for STMR? Obviously a type must never be required for one, but is there a problem with having an explicit assertion for one?

Reply:
I think that depends on the resolution of https://github.com/tc39/proposal-import-assertions/issues/49. If a decision is made there that `js` or `javascript` means Source Text Module Records that would be a check that we add before invoking HostGetSupportedExtraModuleScriptTypes. (I guess we also want to forbid hosts from using values such as js/javascript/ecmascript come to think of it.) Are there plans for a registry?

Reply:
I very much doubt there could be any kind of consensus for a registry; but I’m not up to speed on all the built-in module discussions on that topic.

If the spec prohibits any host-defined type value from representing STMR, but mandates a specific one be supported, that seems fine with me - but then we have to decide how to spell it :-)

Reply:
At some point @littledan created a repository for host coordination I think. Perhaps we can just maintain an informal list there.

Reply:
@ljharb @annevk You're right, it doesn't really work to try to restrict type `js`/`javascript`/etc here.  I've removed this part of the change.

Considering JSON modules, EcmaScript says that if there is `type: "json"` then the import must be a JSON module or fail to load, but EcmaScript takes no position on whether a host could load JSON modules using other values of the `type` assertion in addition to `"json"`. So maybe it's wrong to single out Source Text Modules as needing restrictions like this.

Early thinking about a registry was included in this repo at some point but was [removed](https://github.com/tc39/proposal-import-assertions/pull/48) so as not to draw too much focus from the core of the proposal.  I'm not sure it's been discussed much since. Our best bet may indeed be to maintain an informal list like that (but where?), and ultimately leave such decisions up to the host.

Reply:
I still think you'd want to at least reserve es/ecmascript/js/javascript, perhaps with an assertion.

Reply:
I finally found it, we should document these strings in @littledan's https://github.com/littledan/js-shared-interfaces.